include/neml2/dispatcher/WorkDispatcher.h

```diff
@@ -0,0 +1,253 @@
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+#pragma once
+
+#include <functional>
+#include <future>
+
+#include "neml2/dispatcher/WorkGenerator.h"
+#include "neml2/dispatcher/WorkScheduler.h"
+#include "neml2/misc/error.h"
+#include "neml2/misc/types.h"
+
+// Pre-C++20 workaround for std::type_identity
+// https://en.cppreference.com/w/cpp/types/type_identity
+template <class T>
+struct type_identity
+{
+  using type = T;
+};
+
+namespace neml2
+{
+/**
+ * @brief A simple work dispatcher which dispatches work sequentially without asynchronous execution
+ *
+ * The work dispatcher is responsible for dispatching work generated by a work generator. The
+ * dispatcher also takes care of preprocessing, postprocessing, and reducing the work. In general,
+ * each work dispatch involves four steps:
+ *
+ * 1. Work generation: The work generator generates the next \p n batches of work.
+ * 2. Preprocessing: The dispatcher preprocesses the work.
+ * 3. Do work: The dispatcher dispatches the work to a worker, and the worker complets and returns
+ *    the result.
+ * 4. Postprocessing: The dispatcher postprocesses the result.
+ *
+ * Once all the work has been completed and results have been collected, the dispatcher reduces the
+ * results to obtain the final result.
+ *
+ * @tparam I Input type of the preprocessed work (generated by the generator)
+ * @tparam O Output type of the result returned by the worker
+ * @tparam Of Output type of the final result (after reduction)
+ * @tparam Ip Input type of the work before preprocessing
+ * @tparam Op Output type of the result after postprocessing
+ */
+template <typename I,
+          typename O,
+          typename Of = typename std::vector<O>,
+          typename Ip = typename type_identity<I>::type,
+          typename Op = typename type_identity<O>::type>
+class WorkDispatcher
+{
+public:
+  WorkDispatcher(std::function<O(I &&, torch::Device)> && dispatch)
+    : _dispatch(std::move(dispatch))
+  {
+  }
+
+  WorkDispatcher(std::function<O(I &&, torch::Device)> && dispatch,
+                 std::function<O(std::vector<O> &&)> && reduce)
+    : _dispatch(std::move(dispatch)),
+      _reduce(std::move(reduce))
+  {
+  }
+
+  WorkDispatcher(std::function<O(I &&, torch::Device)> && dispatch,
+                 std::function<Of(std::vector<Op> &&)> && reduce,
+                 std::function<I(Ip &&, torch::Device)> && preprocess,
+                 std::function<Op(O &&)> && postprocess)
+    : _dispatch(std::move(dispatch)),
+      _reduce(std::move(reduce)),
+      _preprocess(std::move(preprocess)),
+      _postprocess(std::move(postprocess))
+  {
+  }
+
+  /// Run the dispatching loop
+  Of run(WorkGenerator<Ip> &, WorkScheduler &) const;
+
+  /// Run the dispatching loop asynchronously
```

Comment:
Perhaps I will understand more when I read the implementation, but doesn't this contradict the object comment?